protocols/kad/src/behaviour.rs

```diff
@@ -2433,14 +2433,17 @@ where
         // Run the periodic provider announcement job.
         if let Some(mut job) = self.add_provider_job.take() {
             let num = usize::min(JOBS_MAX_NEW_QUERIES, jobs_query_capacity);
+            let mut ready_provider_keys: Vec<Key> = Vec::new();
             for _ in 0..num {
                 if let Poll::Ready(r) = job.poll(cx, &mut self.store, now) {
-                    self.start_add_provider(r.key, AddProviderContext::Republish)
-                } else {
-                    break;
```

Comment:
I think a better fix would be to move the `jobs_query_capacity -= ` statement before the break here but use the index of the `for` loop to decrease it by the right amount.